### PR TITLE
Take users to the expense's page after submitting

### PR DESCRIPTION
**Pull request by sudo-kaizen**
This PR aims to fix [issue 2109](https://github.com/opencollective/opencollective/issues/2109)

Following the specs on the issue: 

1. Submitters are redirected to `/{collective_slug}/expenses/{expense_id}` with Router.pushRoute when the expense has been created sucessfully with a special query param to notify the page that the expense has just been created, like `?createSuccess=true`

2.  The message currently in  https://github.com/opencollective/opencollective-frontend/blob/8ff7e6140d849e095ff9403de4bd1052cb531712/pages/createExpense.js#L98  is shown on the page redirected to with the message `(Your expense has been submitted with success...).`

2109_comment_01

### CODE REVIEW

**Comment by sudo-kaizen** [comment]
Hello @Betree, could you please help me with how can I pass the failing CI test?

File: pages/createExpense.js [hunk 1]
```diff
@@ -57,10 +57,12 @@ class CreateExpensePage extends React.Component {
       console.log('>>> createExpense', expense);
       const res = await this.props.createExpense(expense);
       console.log('>>> createExpense res', res);
-      this.setState({
-        showNewExpenseForm: false,
-        expenseCreated: res.data.createExpense,
-      });
+      const expenseCreated = res.data.createExpense;
+      Router.pushRoute(`/${this.props.slug}/expenses/${expenseCreated.id}/?createSuccess=true`);
+      // this.setState({
```
Comment (Betree) [inline pages/createExpense.js]: No need to comment, you can remove these lines directly

File: server/pages.js [hunk 2]
```diff
@@ -43,7 +43,10 @@ pages
   .add('updates', '/:collectiveSlug/updates')
   .add('update', '/:collectiveSlug/updates/:updateSlug')
   .add('createExpense', '/:parentCollectiveSlug?/:type(events)?/:collectiveSlug/expenses/new')
-  .add('expense', '/:parentCollectiveSlug?/:collectiveType(events)?/:collectiveSlug/expenses/:ExpenseId([0-9]+)')
+  .add(
+    'expense',
+    '/:parentCollectiveSlug?/:collectiveType(events)?/:collectiveSlug/expenses/:ExpenseId([0-9]+)/:createSuccess?',
```
Comment (Betree) [inline server/pages.js]: No need to change the route for URL parameters

File: pages/expense.js [hunk 3]
```diff
@@ -35,9 +39,12 @@ class ExpensePage extends React.Component {
   }
 
   render() {
-    const { data, ExpenseId, LoggedInUser } = this.props;
+    const { data, ExpenseId, LoggedInUser, createSuccess } = this.props;
+    const expenseCreated = {
```
Comment (Betree) [inline pages/expense.js]: What's the intent to create this object rather than using `id` directly?

File: pages/expense.js [hunk 4]
```diff
@@ -35,9 +39,12 @@ class ExpensePage extends React.Component {
...
+      id: ExpenseId,
+    };
 
-    if (!data.Collective) return <ErrorPage data={data} />;
+    if (!data.Collective || data.Collective.host) return <ErrorPage data={data} />;
```
Comment (Betree) [inline pages/expense.js]: This returns an error if the collective has a host, which is a problem. Whether the collective has a host or not should not trigger any error.

File: pages/expense.js [hunk 5]
```diff
@@ -89,6 +96,39 @@ class ExpensePage extends React.Component {
                   </Link>
                 </div>
 
+                <Box width={[1, null, 3 / 4]}>
+                  {createSuccess && (
+                    <Box m={3}>
+                      <p className="createSuccess">
```
Comment (Betree) [inline pages/expense.js]: Keeping the old className (expenseCreated) should fix the test.

File: pages/expense.js (hunk 3 above)
Comment (sudo-kaizen) [reply pages/expense.js]: The `ExpenseWithData` component used the same variable name `ExpenseId` for its `id` attribute

File: pages/expense.js (hunk 4 above)
Comment (sudo-kaizen) [reply pages/expense.js]: You mentioned something like that here https://github.com/opencollective/opencollective/issues/2109#issuecomment-530424085

**Comment by sudo-kaizen** [comment]
@Betree Will running `npm run test` locally after making these fixes help ?

**Comment by sudo-kaizen** [comment]
YASS!! @Betree  Thanks much! It's passing now!! 🙌😂😂

File: pages/expense.js (hunk 4 above)
Comment (Betree) [reply pages/expense.js]: Yes, because we shouldn't try to show the host name if the collective has no host. But we must still display the expense info.

**Comment by Betree** [comment]
It works, good job! I'll do a small change today to make it consistent with design

**Comment by sudo-kaizen** [comment]
Hello @Betree is there anything I can help you with here?

**Comment by Betree** [comment]
@sudo-kaizen yes, can you allow push from maintainers on your branch? See https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by sudo-kaizen** [comment]
> @sudo-kaizen yes, can you allow push from maintainers on your branch? See https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

It is enabled already

**Comment by Betree** [comment]
Ok for some reason it's blocking me from pushing, I'll re-create the PR

**Comment by Betree** [comment]
Moved to https://github.com/opencollective/opencollective-frontend/pull/2651

**Pull request closed without merging**